Make "Multi selects focus the first selected option when opened"

Samic8 opened the pull request:
Hi @mitchbne 

Like I mentioned in #2 I want to do lots of small PRs to get early feedback from you. Have a look over this and see what you think about the approach. Note that this is going to a staging "Multiselect" branch.

I'm not entirely convinced about having to check if `this.props.value` is an array or not everywhere. Alternatively, we could have a `values` prop that needs to be used with the Multiselect version.

Samic8 committed:
Make "Multi selects focus the first selected option when opened"

Samic8 committed:
Handle default focus item

Samic8 committed:
Fix naming typo

mitchbne commented on src/Listbox.tsx:
```diff
@@ -101,21 +97,34 @@ export class Listbox extends Component<Props, State> {
+    // in the list if no item is selected.
+    // https://www.w3.org/TR/wai-aria-practices/#listbox_kbd_interaction
+    if (!activeValue){ return this.state.values[0] }
+    return Array.isArray(activeValue) ? activeValue[0] : activeValue
```
Can you please extract out something like `isMultiSelect` (there seems to just be a lot of `isArray()` repeated over and over), as perhaps a state variable on `Listbox`, and pass that through the ListboxContext?

Samic8 replied:
Yep sounds good

Samic8 replied:
What do you think about having a `values` prop for the mutliselect version? @mitchbne 
That way we don't have to check if it's an array or not everywhere.

We could also error early in the `Listbox` component to prevent people from using the wrong props
```
if (this.props.multiselect && this.props.value) {
    throw new Error(`Use the "values" prop for multiselect`)
}

if (!this.props.multiselect && this.props.values) {
   throw new Error(`Use the "value" prop for non-multiselect`)
}
```

Samic8 replied:
I think I figured out a way to do this without errors in typescript with a union type
```
interface SharedProps {
  children: React.ReactNode,
  className?: string,
  onChange: (value: string | string[] | null) => void
}

interface SingleProps extends SharedProps {
  multiselect: false;
  value: string | null,
}

interface MultiProps extends SharedProps {
  multiselect: true;
  values: string[];
}

type Props = SingleProps | MultiProps
```

Samic8 replied:
I've pushed some changes to do it the union type way above ^

Keen to hear your feedback! Thanks

mitchbne replied:
I would additionally add-in `never`.

```
interface SharedProps {
  children: React.ReactNode
  className?: string
  onChange: (value: string | string[] | null) => void
}

interface SingleProps extends SharedProps {
  multiselect: false
  value: string | null
  values: never
}

interface MultiProps extends SharedProps {
  multiselect: true
  values: string[]
  value: never
}

type Props = SingleProps | MultiProps
```

Samic8 replied:
Sounds good. I added that in: https://github.com/mitchbne/react-listbox/pull/5/commits/f9f2f0caaefa13f092950665acc16524b4d6186f

Samic8 replied:
Actually this seems to create a TS error. Looking into why.

Samic8 replied:
I removed it again. Seems it was expecting the property to be defined even though it was never. If you figure it out let me know. Thanks!

Samic8 commented on src/Listbox.tsx:
```diff
@@ -101,21 +97,34 @@ export class Listbox extends Component<Props, State> {
     this.setState({ activeItem: value }, () => {
       if (value === null){ return }
-      this.state.listboxListRef?.children[this.state.values.indexOf(this.state.activeItem)].scrollIntoView({ block: "nearest" })
+      this.state.listboxListRef?.children[this.state.values.indexOf(this.getFocusValue())].scrollIntoView({ block: "nearest" })
```
🤦  This should use value instead of the `getFocusValue` function.

I'll rename `getFocusValue` to `getDefaultFocusValue` to make it clearer what that function does.

Samic8 replied:
Done now

Samic8 committed:
Use union type for mutli/single select props

Samic8 committed:
Make multiselect prop optional for single select version

Samic8 committed:
Add tests for closing the listbox

Samic8 committed:
Add "never" to typescript properties

Samic8 committed:
Remove never

Samic8 committed:
Make seperate onChange types

Samic8 committed:
Add storybook example

Samic8 committed:
Add sorting for onChange

Samic8 commented:
This PR is ready to review again @mitchbne. I'm going to branch off this and start work on the keyboard shortcuts.

Samic8 committed:
Add shift key shortcuts for arrows

Samic8 commented:
Found a bug. I didn't test/implement unselecting values.

Samic8 committed:
Handle unselecting values

Samic8 commented:
> Found a bug. I didn't test/implement unselecting values.

Fixed in my last commit

Samic8 committed:
Handle contiguous item selecting

Samic8 committed:
Merge branch 'multiselect-start' into multiselect-shortcuts

Samic8 committed:
Add range select shortcuts

Samic8 committed:
Add more shortcuts for multiselect

Control + Shift + Home: Selects the focused option and all options up to the first option. Optionally, moves focus to the first option.
Control + Shift + End: Selects the focused option and all options down to the last option. Optionally, moves focus to the last option.
Control + A: Selects all options in the list. Optionally, if all options are selected, it may also unselect all options.

Samic8 committed:
Check if multiselect before selecting many

mitchbne committed:
Merge pull request #7 from mitchbne/multiselect-shortcuts

mitchbne committed:
Merge pull request #8 from mitchbne/more-shortcuts
